pandas/core/algorithms.py

```diff
@@ -579,6 +572,34 @@ def f(c, v):
             f = lambda a, b: np.isin(a, b).ravel()
 
     else:
+        # Fast-path: handle integer-kind mixes without upcasting to float64.
+        if (
+            values.dtype.kind in "iu"
+            and comps_array.dtype.kind in "iu"
+            # Only apply fast-path for 64-bit integer widths to avoid
```

Comment:
Can you expand on this comment? I don't fully understand what the surprising behavior is, and I'm worried its a red herring